idaes/apps/uncertainty_propagation/uncertainties.py

```diff
@@ -0,0 +1,230 @@
+        fxxp.append(fxxp_tmp)
+    fxxp = np.array(fxxp)
+    # calculate error propagation df/dp*cov_p*df/dp + (df/dx*dx/dp)*cov_p*(df/dx*dx/dp)
+    # objective function: 
+    for cc in list(model.component_data_objects(Objective,
+                                            active=True,
+                                            descend_into=True)):
+        # save gradient of the objective with respect to only theta_names
+        if any(ele in str(cc.expr) for ele in theta_names):
+            gradient_f_theta = gradient_f[[int(line_dic[v]) -1 for v in theta_names]]
+    propagation_f = {}
+    # calculate error propagation of the objective fuction
+    if 'gradient_f_theta' in locals():
+        propagation_f['objective'] = np.dot(gradient_f_theta,np.dot(cov,np.transpose(gradient_f_theta))) + np.dot(fxxp,np.dot(cov,np.transpose(fxxp)))
+    else:
+        propagation_f['objective'] = np.dot(fxxp,np.dot(cov,np.transpose(fxxp)))
+    # constraints: 
+    # calculate error propagation dc/dp*cov_p*dc/dp + (dc/dx*dx/dp)*cov_p*(dc/dx*dx/dp)
+    num_constraints = len(list(model.component_data_objects(Constraint,
+                                                            active=True,
+                                                            descend_into=True))) 
+    if num_constraints > 0:
+        # save constraints into gradient_cc only if constraints includes theta 
+        # gradient_c[i] is a sparse matrix with column number, row number, and value
+        gradient_cc = []
+        for i in range(gradient_c.shape[0]):
+            if gradient_c[i,[0]] in line_dic.values():
+                gradient_cc.append(gradient_c[i])
+        gradient_cc = np.array(gradient_cc)
+        # save unique constraints numbers that contain theta, index starts from 1. 
+        constriant_number = list(set(gradient_cc[:,[1]].flatten().astype(int)))       
+        # convert sparse matrix to dense matrix with (value, (row number, column number))
+        I,J,V =  gradient_cc[:,[1]].flatten().astype(int)-1, gradient_cc[:,[0]].flatten().astype(int)-1,  gradient_cc[:,[2]].flatten()
+        gradient_cc = sparse.coo_matrix((V,(I,J))).todense()
+        gradient_dic = {}
+        propagation_c = {}
+        # calculate error propagation of constraints
+        for r in range(1,num_constraints+1):
+            if r in constriant_number:
+                cxxp = []
+                for i in p_list:
+                    cxxp_tmp = 0
+                    for j in x_list:
+                        cxxp_tmp = 0
+                        # we took the gradients from a sparse matrix
+                        if 'd(c'+str(r)+')/d('+j+')' in gradient_c_dic.keys():
+                            gradient_cx = gradient_c_dic['d(c'+str(r)+')/d('+j+')']
+                        else:
+                            gradient_cx = 0
+                        cxxp_tmp = cxxp_tmp + gradient_cx*dsdp_dic['d('+j+')/d('+i+')']
+                    cxxp.append(cxxp_tmp)
```

Comment:
Add more comments please.

Reply:
From [Line `cxxp = [ ]`](https://github.com/IDAES/idaes-pse/blob/0854fab43a0e50da7860682718bdd3864fbabd9f/idaes/apps/uncertainty_propagation/uncertainties.py#L175) to[ Line `cxxp = np.array(cxxp)` ](https://github.com/IDAES/idaes-pse/blob/0854fab43a0e50da7860682718bdd3864fbabd9f/idaes/apps/uncertainty_propagation/uncertainties.py#L187) are redundant codes. I removed them.:set